[Merged by Bors] - feat(AlgebraicTopology/SimplicialObject): define simplicial homotopy

foderm opened the pull request:
This file defines the notion of a combinatorial simplicial homotopy between two morphisms of simplicial objects.

A first step towards proving homotopy invariance of simplicial homotopy. See [Zulip](https://leanprover.zulipchat.com/#narrow/channel/335062-homology/topic/Homotopy.20invariance.20of.20singular.20homology/with/562221688). 

---
<!-- Your PR title will become the first line of the commit message.

In this box, the text above the `---` (if not empty) will be appended
to the commit message, and can be used to give additional context or
details. Please leave a blank newline before the `---`, otherwise GitHub
will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

When merging, all the commits will be squashed into a single commit
listing all co-authors.

Co-authors in the squash commit are gathered from two sources:

First, all authors of commits to this PR branch are included. Thus,
one way to add co-authors is to include at least one commit authored by
each co-author among the commits in the pull request. If necessary, you
may create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

Second, co-authors can also be listed in lines at the very bottom of
the commit message (that is, directly before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines
at the bottom of the commit message (before the `---`, and also before
any "Co-authored-by" lines) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

joelriou commented on Mathlib/AlgebraicTopology/SimplicialObject/SimplicialHomotopy.lean:
```diff
@@ -0,0 +1,206 @@
+/-- A simplicial homotopy between morphisms `f g : X ⟶ Y` of simplicial objects
+consists of a family of morphisms `h n i : X _⦋n⦌ ⟶ Y _⦋n+1⦌` for `i : Fin (n + 1)`,
+satisfying compatibility conditions with respect to face and degeneracy maps -/
+structure SimplicialHomotopy
```
```suggestion
@[ext]
structure SimplicialHomotopy
```
Doing this make your definition of the `ext` lemma below unnecessary.

joelriou commented on Mathlib/AlgebraicTopology/SimplicialObject/SimplicialHomotopy.lean:
```diff
@@ -0,0 +1,206 @@
+
+/-- The constant homotopy from `f` to `f`. -/
+@[simps]
+def refl (f : X ⟶ Y) : SimplicialHomotopy f f where
```
After the changes suggested above, this may become:
```lean
/-- The constant homotopy from `f` to `f`. -/
@[simps]
def refl (f : X ⟶ Y) : SimplicialHomotopy f f where
  h i := X.σ i ≫ f.app _
  h_zero_comp_δ_zero n := by
    have := Y.δ_comp_σ_self (n := n) (i := 0)
    dsimp at this
    simp [this]
  h_last_comp_δ_last n := by
    have := Y.δ_comp_σ_succ (i := Fin.last n)
    dsimp at this
    simp [this]
  h_succ_comp_δ_castSucc_of_lt i j hij := by simp [Y.δ_comp_σ_of_le hij]
  h_succ_comp_δ_castSucc_succ j := by
    have h₁ := Y.δ_comp_σ_self (i := j.succ)
    have h₂ := Y.δ_comp_σ_succ (i := j.castSucc)
    dsimp at h₁ h₂
    simp [h₁, h₂]
  h_castSucc_comp_δ_succ_of_lt i j hij := by simp [Y.δ_comp_σ_of_gt hij]
  h_comp_σ_castSucc_of_le i j hij := by simp [Y.σ_comp_σ hij]
  h_comp_σ_succ_of_lt i j hij := by simp [Y.σ_comp_σ hij]
```
The main point is that we already have lemmas translating of the simplicial relations as equalities of compositions of faces/degeneracies acting on a simplicial object.

joelriou commented:
In order to be sure the equations are correct, it would be safe to wait for the formalization of the chain homotopy on alternating face map complexes that is induced by a simplicial homotopy (in case `C` is a preaddive category).

foderm committed:
Added postcomposition with a functor to SimplicialHomotopy
Constructed the induced chain homotopy on alternatingFaceMapComplex and deduce equality on homology.
And deduced the simplicial-step homotopy invariance statement for singular chain complexes/homology.

foderm committed:
run lake exe mk_all

foderm committed:
conflicts

foderm committed:
module

foderm committed:
public section

foderm committed:
added docstring

foderm committed:
added docstring

foderm committed:
switched homotopy endpoints and added some simp lemmas

foderm committed:
docstrings

foderm committed:
new hom definition

foderm committed:
removed SimplicialHomotopy variable in the lemmas

joelriou commented on Mathlib/AlgebraicTopology/SimplicialObject/SimplicialHomotopy.lean:
```diff
@@ -0,0 +1,183 @@
+    {X Y : SimplicialObject C} (f g : X ⟶ Y) where
+  /-- Basic data: `h i : Xₙ ⟶ Yₙ₊₁` for `i = 0..n`. -/
+  h {n : ℕ} (i : Fin (n + 1)) : (X _⦋n⦌ ⟶ Y _⦋n+1⦌)
+  /-- Endpoint `d₀ h₀ = g`. -/
```
```suggestion
```

joelriou commented on Mathlib/AlgebraicTopology/SimplicialObject/SimplicialHomotopy.lean:
```diff
@@ -0,0 +1,183 @@
+  h_comp_σ_castSucc_of_le {n : ℕ} (i j : Fin (n + 1)) (hij : i ≤ j) :
+    h j ≫ Y.σ i.castSucc = X.σ i ≫ h j.succ
+  /- nlab: `sᵢ hⱼ = hⱼ s_{i'-1}` if i' > j, let i' = i + 1 -/
+  /-- `s_{i+1} hⱼ = hⱼ sᵢ` if i + 1 > j -/
```
```suggestion
```

joelriou commented on Mathlib/AlgebraicTopology/SimplicialObject/SimplicialHomotopy.lean:
```diff
@@ -0,0 +1,183 @@
+
+/-- Postcompose a simplicial homotopy with a functor `F : C ⥤ D`. -/
+@[simps]
+def whiskerRight (H : SimplicialHomotopy f g) :
```
Putting the functor as second parameter will allow dot notation `H.whiskerRight F`.
```suggestion
/-- Postcompose a simplicial homotopy with a functor `F : C ⥤ D`. -/
@[simps]
def whiskerRight (H : SimplicialHomotopy f g) {D : Type u'} [Category.{v'} D] (F : C ⥤ D) :
```
(Note that this will probably break things in other files.)

joelriou commented on Mathlib/AlgebraicTopology/SimplicialObject/ChainHomotopy.lean:
```diff
@@ -0,0 +1,316 @@
+This is the alternating sum `H_n = ∑ i : Fin (n+1), (-1)^i • H.h i`.
+-/
+def chainHomotopyComponent (H : SimplicialHomotopy f g) (n : ℕ) :
+    X _⦋n⦌ ⟶ Y _⦋n+1⦌ := ∑ i : Fin (n+1), ( (-1 : ℤ) ^ (i : ℕ) : ℤ ) • H.h i
```
This is unused.

joelriou commented on Mathlib/AlgebraicTopology/SimplicialObject/ChainHomotopy.lean:
```diff
@@ -0,0 +1,316 @@
+    ((alternatingFaceMapComplex C).map g).f 0 =
+    prevD 0 (hom H) + ((alternatingFaceMapComplex C).map f).f 0 := by
+  rw [prevD_eq (hom H) (j := 0) (j' := 1) (by simp)]
+  simp [← H.h_last_comp_δ_last 0]
```
```suggestion
private lemma comm_zero :
    letI d : Y _⦋1⦌ ⟶ Y _⦋0⦌ := ((alternatingFaceMapComplex C).obj Y).d 1 0
    f.app (op ⦋0⦌) = hom H 0 1 ≫ d + g.app (op ⦋0⦌) := by
  simp [← H.h_last_comp_δ_last 0]
```

foderm committed:
Update Mathlib/AlgebraicTopology/SimplicialObject/SimplicialHomotopy.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

foderm committed:
Update Mathlib/AlgebraicTopology/SimplicialObject/SimplicialHomotopy.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

foderm committed:
Update Mathlib/AlgebraicTopology/SimplicialObject/SimplicialHomotopy.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

foderm committed:
Update Mathlib/AlgebraicTopology/SimplicialObject/SimplicialHomotopy.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

foderm committed:
Update Mathlib/AlgebraicTopology/SimplicialObject/SimplicialHomotopy.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

foderm committed:
Update Mathlib/AlgebraicTopology/SimplicialObject/SimplicialHomotopy.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

foderm committed:
Update Mathlib/AlgebraicTopology/SimplicialObject/SimplicialHomotopy.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

foderm committed:
Update Mathlib/AlgebraicTopology/SimplicialObject/SimplicialHomotopy.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

foderm committed:
Update Mathlib/AlgebraicTopology/SimplicialObject/SimplicialHomotopy.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

foderm committed:
Update Mathlib/AlgebraicTopology/SimplicialObject/SimplicialHomotopy.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

foderm committed:
namespace changes

foderm committed:
Update Mathlib/AlgebraicTopology/SimplicialObject/ChainHomotopy.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

foderm committed:
Update Mathlib/AlgebraicTopology/SimplicialObject/ChainHomotopy.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

foderm committed:
shorter proof from joelriou

joelriou commented on Mathlib/AlgebraicTopology/SimplicialObject/ChainHomotopy.lean:
```diff
@@ -0,0 +1,187 @@
+      rw [prevD_eq (j' := 1) (w := by simp), dNext_eq_zero _ _ (by simp), zero_add]
+      simp [ToChainHomotopy.comm_zero H]
+    | succ n =>
+      rw [dNext_eq (i' := n) (w := by simp), prevD_eq (j' := n + 1 + 1) (w := by simp)]
```
```suggestion
      rw [dNext_eq (i' := n) (w := by simp), prevD_eq (j' := n + 2) (w := by simp)]
```

foderm committed:
renamed homotopy file

foderm committed:
Update Mathlib/AlgebraicTopology/SimplicialObject/ChainHomotopy.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

foderm committed:
Update Mathlib/AlgebraicTopology/SimplicialObject/ChainHomotopy.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

foderm committed:
CI fail

foderm committed:
CI fail

foderm committed:
pushed the wrong file

foderm committed:
CI fail

foderm committed:
ok now it should be correct

joelriou commented on Mathlib/AlgebraicTopology/SimplicialObject/Homotopy.lean:
```diff
@@ -0,0 +1,167 @@
+
+attribute [reassoc (attr := simp)]
+  h_zero_comp_δ_zero h_last_comp_δ_last h_succ_comp_δ_castSucc_of_lt h_succ_comp_δ_castSucc_succ
+  h_castSucc_comp_δ_succ_of_lt h_comp_σ_castSucc_of_le h_comp_σ_succ_of_lt
```
It seems unclear to me that `h_succ_comp_δ_castSucc_succ` should be a simp lemma:
```suggestion
attribute [reassoc (attr := simp)]
  h_zero_comp_δ_zero h_last_comp_δ_last h_succ_comp_δ_castSucc_of_lt
  h_castSucc_comp_δ_succ_of_lt h_comp_σ_castSucc_of_le h_comp_σ_succ_of_lt

attribute [reassoc] h_succ_comp_δ_castSucc_succ
```

joelriou commented:
Could you merge with the latest `master`?
If proofs break for some lemmas, you may try inserting this line :
```lean
set_option backward.isDefEq.respectTransparency false in
lemma ...
```

foderm committed:
Update Mathlib/AlgebraicTopology/SimplicialObject/ChainHomotopy.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

foderm committed:
Update Mathlib/AlgebraicTopology/SimplicialObject/Homotopy.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

foderm committed:
Update Mathlib/AlgebraicTopology/SimplicialObject/Homotopy.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

foderm committed:
Update Mathlib/AlgebraicTopology/SimplicialObject/Homotopy.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

foderm committed:
moved Fintype lemma

joelriou committed:
Merge branch 'master' into Brouwer

joelriou committed:
fix

foderm committed:
removed simp attribute from h_succ_comp_δ_castSucc_succ

foderm committed:
Merge branch 'Brouwer' of https://github.com/foderm/mathlib4 into Brouwer
